### PR TITLE
improve favicon caching

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1183977322430184
Tech Design URL:
CC:

**Description**:

Improves favicon caching as follows:

* tabs have their own cache
* bookmarks, favorites and fireproofed (collectively referred to as bookmarks) site have their own cache
* items are only added to the caches at the appropriate time
  * tabs when the user browses 
  * bookmarks when they are added or launched - taking from the tabs cache first, if available
* items expire after a week but the cache remains in place and is used until user interacts with the image
* uses the default web view user agent for requests instead of networking default
* file names a salted hash of the URL's domain, ie hash = sha256("DDGSalt:" + domain)

**Steps to test this PR**:

***Caching***:
1. add a site to favorites - news.ycombinator.com is a good one
1. ensure image is loaded and looks good then clear all tabs and data
1. note location of cache directory for bookmarks (logged to console, filter on `favicons`)
1. kill the app
1. open the cache directory in terminal and touch the cached image to a time 5 minutes from now using `touch -mt YYYYMMDDhhmm filename`
1. launch the app - the image should appear and look fine, the modified time on the cached image should be unchanged
1. tap the favorite - the website should load and the modified time on the cached image should change to a week from now

***Clearing tabs***:
1. browse to a bunch of sites
1. check the tabs cache and there will be a representative number of images
1. clear all tabs and data
1. all the images will be gone

***Check fallbacks***:
1. Open google.com (does not have apple-touch-icon.png) - check tab switcher, icon should appear and be cached

***Removing bookmark cache items***:
1. Favorite a site, check entry appears in cache
1. Fireproof same site, should show icon in fireproof
1. Remove Fireproof, cache entry still present
1. Remove Favorite, cache entry is gone
1. Repeat with various combinations of bookmarks, favorites, and fireproof

***Network requests***:
1. Check that network requests have the web view user agent associated with them, rather than the default networking user agent.

**Device Testing**:

* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 10
* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/AppUrls.swift [hunk 1]
```diff
@@ -134,9 +119,12 @@ public struct AppUrls {
             .addParam(name: Param.setAtb, value: setAtb)
     }
 
+    public static func isDuckDuckGo(domain: String?) -> Bool {
+        return domain == AppUrls.ddgDomain || domain?.hasSuffix(".\(AppUrls.ddgDomain)") ?? false
```
Comment (bwaresiak) [inline Core/AppUrls.swift]: To consider if possible: use `isPart(ofDomain:)`

File: Core/FaviconSourcesProvider.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    
+    func additionalSources(forDomain domain: String) -> [String] {
+        return [
+            imageSource(forDomain: domain, imageName: .favicon, secure: true),
+            imageSource(forDomain: domain, imageName: .favicon, secure: false)
+        ].compactMap { $0 }
+    }
+    
+    private func imageSource(forDomain domain: String, imageName: ImageNames, secure: Bool) -> String? {
+        return URL(string: (secure ? "https" : "http") + "://" + domain + "/" + imageName.rawValue)?.absoluteString
```
Comment (bwaresiak) [inline Core/FaviconSourcesProvider.swift]: Why return String instead of URL?

File: DuckDuckGo/AppDelegate.swift [hunk 3]
```diff
@@ -54,7 +55,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
             window?.rootViewController = UIStoryboard.init(name: "LaunchScreen", bundle: nil).instantiateInitialViewController()
             return true
         }
+
+        // Do this ASAP or it crashes when initialised off the main thread
+        _ = UserAgentManager.shared
```
Comment (bwaresiak) [inline DuckDuckGo/AppDelegate.swift:59]: I'm concerned about possible Background start flow - maybe keep this confined (along with the migration) to `onApplicationLaunch`?

File: DuckDuckGo/Favicons.swift [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        case bookmarks
+
+        func create() -> ImageCache {
+
+            let imageCache: ImageCache
+
+            switch self {
+            case .bookmarks:
+                let groupName = BookmarkUserDefaults.Constants.groupName
+                let sharedLocation = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupName)
```
Comment (bwaresiak) [inline DuckDuckGo/Favicons.swift]: I this backed up to iCloud? I'd keep this in subdirectory and exclude it form backups.

File: DuckDuckGo/Favicons.swift [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                os_log("favicons bookmarks location %s", log: generalLog, type: .debug, sharedLocation?.absoluteString ?? "<none>")
+                imageCache = (try? ImageCache(name: self.rawValue, cacheDirectoryURL: sharedLocation))
+                                ?? ImageCache(name: self.rawValue)
+                            
+            case .tabs:
+                let location = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
```
Comment (bwaresiak) [inline DuckDuckGo/Favicons.swift]: What happens if that is cleared? Maybe we should move this to application support dir, like we did with tab previews.

File: DuckDuckGo/BookmarksManager.swift [hunk 6]
```diff
@@ -107,15 +108,31 @@ class BookmarksManager {
...
     func deleteBookmark(at index: Int) {
+        let link = bookmark(atIndex: index)
         var bookmarks = dataStore.bookmarks
         bookmarks.remove(at: index)
         dataStore.bookmarks = bookmarks
+        removeFavicon(forLink: link)
     }
 
     func deleteFavorite(at index: Int) {
+        let link = favorite(atIndex: index)
```
Comment (bwaresiak) [inline DuckDuckGo/BookmarksManager.swift:117]: What about edit action?

File: DuckDuckGoTests/FaviconSourcesProviderTests.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
+//
+//  FaviconSourcesProviderTests.swift
+//  Core
+//
+//  Created by Chris Brind on 15/07/2020.
+//  Copyright © 2020 DuckDuckGo. All rights reserved.
+//
```
Comment (bwaresiak) [inline DuckDuckGoTests/FaviconSourcesProviderTests.swift:6]: Copyright? :)

File: Core/AppUrls.swift (hunk 1 above)
Comment (brindy) [reply Core/AppUrls.swift]: As discussed, optimisation to avoid creating AppUrls instance

File: Core/FaviconSourcesProvider.swift (hunk 2 above)
Comment (brindy) [reply Core/FaviconSourcesProvider.swift]: Good point, changed

File: DuckDuckGo/AppDelegate.swift (hunk 3 above)
Comment (brindy) [reply DuckDuckGo/AppDelegate.swift:59]: Good call, thanks

File: DuckDuckGo/Favicons.swift (hunk 4 above)
Comment (brindy) [reply DuckDuckGo/Favicons.swift]: Added flags to prevent backup.

File: DuckDuckGo/Favicons.swift (hunk 5 above)
Comment (brindy) [reply DuckDuckGo/Favicons.swift]: Done, thanks!

File: DuckDuckGo/BookmarksManager.swift (hunk 6 above)
Comment (brindy) [reply DuckDuckGo/BookmarksManager.swift:117]: Done, thanks!

File: DuckDuckGoTests/FaviconSourcesProviderTests.swift (hunk 7 above)
Comment (brindy) [reply DuckDuckGoTests/FaviconSourcesProviderTests.swift:6]: It's mine, all mine!

ok ok :)

File: DuckDuckGo/Favicons.swift [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+            return cache
+        }
+        
+        private func createCacheInDesiredLocation() -> ImageCache? {
+            
+            guard var url = baseCacheURL()?.appendingPathComponent(Constants.faviconsFolderName) else { return nil }
+            
+            if !FileManager.default.fileExists(atPath: url.path) {
+                try? FileManager.default.createDirectory(at: url,
+                                        withIntermediateDirectories: false,
```
Comment (bwaresiak) [inline DuckDuckGo/Favicons.swift]: I'd go with 'true' here, as Application Support directory may not exist.

File: DuckDuckGo/Favicons.swift [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+import os
+
+public class Favicons {
+
+    public struct Constants {
+
+        public static let standardPlaceHolder = UIImage(named: "GlobeSmall")
+        
+        static let salt = "DDGSalt:"
+        static let faviconsFolderName = "favicons"
```
Comment (bwaresiak) [inline DuckDuckGo/Favicons.swift]: Just a nitpick, but I'd start folder name with capital letter. :)

**Comment by brindy** [comment]
@bwaresiak heads up I've made a bit of a change - I was close to this solution the other day but ran out of time, then I realised this morning where I was falling short and decided to go with it.  Essentially a UserScript now drives the population of the tabs cache and the bookmarks favicon will take from the tabs cache if available, or download otherwise.  The result is much better images and improved UX.  Let me know what you think - happy to zoom about it later.  Thanks!

File: DuckDuckGo/Favicons.swift (hunk 8 above)
Comment (brindy) [reply DuckDuckGo/Favicons.swift]: Will change, but FYI I copied this from the tab previews example...

Comment (brindy) [reply DuckDuckGo/Favicons.swift]: Ah, I see - you explicitly create the Application Support directory first.  Got it.

Comment (bwaresiak) [reply DuckDuckGo/Favicons.swift]: I know, but tab previews explicitly create support dir up front before creating the nested one - that could probably be simplified but I wouldn't touch what ain't broken. :)

**Review by bwaresiak — APPROVED** [review]
Great job!

**Comment by brindy** [comment]
Thanks @bwaresiak !

**Pull request merged**